### ISSUE TITLE
Add line-ending detection to linter

### ISSUE THREAD

**Issue by bollwyvl** [conda-forge/conda-smithy#1446]
Seems like a bad `meta.yaml` with CRLF gummed up the works for a bunch of folks on `staged-recipes`... it seems reasonable to detect and warn about line endings.

**Comment by bollwyvl** [comment]
Imagining something like:
```diff
diff --git a/conda_smithy/lint_recipe.py b/conda_smithy/lint_recipe.py
index 1dc271d..83eb043 100644
--- a/conda_smithy/lint_recipe.py
+++ b/conda_smithy/lint_recipe.py
@@ -527,6 +527,15 @@ def lintify(meta, recipe_dir=None, conda_forge=False):
                 "requirement in the future."
             )
 
+    # 26: meta.yaml should not contain windows line-endings
+    if recipe_dir:
+        with io.open(meta_fname, "rt") as fh:
+            if "\r\n" in fh.read():
+                lints.append(
+                    "Windows line endings (CRLF) found in meta.yaml. Please "
+                    "ensure only Unix line endings (LF) are used."
+                )
+
     # hints
     # 1: suggest pip
     if "script" in build_section:
```

- could actually check _all_ yaml
- could hint that they are all valid `utf-8` file, as the local flavor yaml might "helpfully" be working around other encodings that might not _always_ work... `can't decode character at` is one of my least favorite messages :imp:

**Comment by BastianZim** [comment]
action that failed: https://github.com/conda-forge/staged-recipes/runs/1816196440?check_suite_focus=true#step:4:810

The PR with the respective file: https://github.com/conda-forge/staged-recipes/pull/13853

**Comment by BastianZim** [comment]
Just to add, I tried the code on the original file and didn't receive an error. Seems like some platforms automatically correct this so a neutral platform to test it would be great.

**Comment by bollwyvl** [comment]
Nice. It would definitely need a test, and git.autocrlf doesn't help
matters for the specificity...

**Comment by beckermr** [comment]
I have massive levels of confusion about line endings in my head. 

1. The python docs say this

```
In text mode (the default, or when 't' is included in the mode argument), the contents of the file are returned as str, the bytes having been first decoded using a platform-dependent encoding or using the specified encoding if given.
```

See https://docs.python.org/3/library/functions.html#open.

We run the linter on linux. Is the `\r\t` preserved? 

2. Also, the linter clones and checks out the code. The issue we are seeing is that git is changing the file on checkout due to the line endings. I suspect when the linter checks out the code, the same thing will happen and the check above won't be triggered. Will we need to change how the linter checks out the code to make this work?

**Comment by beckermr** [comment]
> Just to add, I tried the code on the original file and didn't receive an error. Seems like some environments automatically correct this so a neutral environment to test it would be great.

@BastianZim I think this is the problem in item 2 I wrote above. Git is making the correction for us.

**Comment by BastianZim** [comment]
> Git is making the correction for us.

@beckermr Yeah agreed. Question would be if we can find a flag that prevents that or otherwise the raw file would have to be downloaded (Probably not the best solution).

**Comment by beckermr** [comment]
I think we want `git config --global core.autocrlf false` maybe?

**Comment by beckermr** [comment]
I found this random blog posty thing: https://jessitron.com/2019/11/11/line-endings-in-git/

**Comment by BastianZim** [comment]
That should fix it. Then the only problem is still your first point. Maybe the linter could open it in binary as no decoding is being doen there

**Comment by BastianZim** [comment]
Found another thing: https://docs.python.org/3/library/functions.html#open-newline-parameter

> When reading input from the stream, if newline is None, universal newlines mode is enabled. Lines in the input can end in '\n', '\r', or '\r\n', and these are translated into '\n' before being returned to the caller. If it is '', universal newlines mode is enabled, but line endings are returned to the caller untranslated. If it has any of the other legal values, input lines are only terminated by the given string, and the line ending is returned to the caller untranslated.

So setting the option `''`  should give us the correct output, if I understand this correctly.

**Comment by isuruf** [comment]
Any solution based on `git config --global core.autocrlf false` is not going to work as some contributors just upload using github UI which doesn't care about line ending settings.

**Comment by BastianZim** [comment]
It would be when the linter clones and checks out the code so wouldn't that be separate from the upload? Local linting might not work but cloud-based one should.

**Comment by beckermr** [comment]
Yes. This is just for when the linter checks out the code so that it can find the offending line endings without git getting in the way.

**Comment by beckermr** [comment]
That change would have to go in the webservices repo.

**Comment by BastianZim** [comment]
Ok then from how I understand it, the change should be:

1. Add `git config --global core.autocrlf false` to the webservices
2. Include @bollwyvl's patch but make sure to open the file with `newline=''`

**Comment by BastianZim** [comment]
Also, it might make sense to add an explanation on how to fix the newlines, or is that too verbose?

**Comment by BastianZim** [comment]
Another possible solution for the git part: How about setting it in _.gitattributes_? 

Explanation is here: https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings

This could also be added to the templates which means that it would also work locally.
